File: Updog.Client/src/post/common/post-mapper.ts
import { Mapper } from '@/core/mapper';
import { Post } from './post';
import { UserMapper } from '@/user/common/user-mapper';

/**
 * Mapper to convert a post into it's entity form from a raw
 * object literal.
 */
export class PostMapper implements Mapper<{ [key: string]: any }, Post> {
    /**
     * Create a new post mapper.
     * @param userMapper The user mapper for converting user entities.
     */
    constructor(private userMapper: UserMapper) {}

    /**
     * Convert an object literal into it's post entity.
     * @param source The raw object literal.
     */
    public map(source: { [key: string]: any }): Post {
        if (typeof source.id !== 'number') {
            throw new TypeError('id must be an integer');
        }

        if (typeof source.type !== 'number') {
            throw new TypeError('type must be an integer');
        }

        if (typeof source.title !== 'string') {
            throw new TypeError('title must be a string');
        }

        if (typeof source.body !== 'string') {
            throw new TypeError('body must be a string');
        }

        if (typeof source.user !== 'object') {
            throw new TypeError('user must be an object');
        }

        if (typeof source.creationDate !== 'string') {
            throw new TypeError('creationDate must be a date string');
        }

        if (typeof source.commentCount !== 'number') {
            throw new TypeError('commentCount must be an integer');
        }

        const user = this.userMapper.map(source.user);
        return new Post(
            source.id,
            source.type,
            source.title,
            source.body,
            user,
            new Date(source.creationDate),
            source.commentCount,
            source.wasUpdated,
            source.wasDeleted
        );
    }
}

File: Updog.Persistance/Post/PostMapper.cs
using System;
using System.Dynamic;
using Updog.Application;
using Updog.Domain;

namespace Updog.Persistance {
    public interface IPostMapper : IReversableMapper<PostRecord, Post> { }

    /// <summary>
    /// Mapper to convert a post record into it's entity and back.
    /// </summary>
    public sealed class PostMapper : IPostMapper {
        #region Publics
        /// <summary>
        /// Convert the record into it's entity form.
        /// </summary>
        /// <param name="source">The source record.</param>
        /// <returns>The entity.</returns>
        public Post Map(PostRecord source) => new Post(
            source.Id,
            source.UserId,
            source.SpaceId,
            source.Type,
            source.Title,
            source.Body,
            source.CreationDate,
            source.CommentCount,
            new VoteStats(source.Upvotes, source.Downvotes),
            source.WasUpdated,
            source.WasDeleted
        );

        /// <summary>
        /// Convert the post back into it's record form.
        /// </summary>
        /// <param name="destination">The entity/</param>
        /// <returns>The rebuilt record.</returns>
        public PostRecord Reverse(Post destination) => new PostRecord {
            Id = destination.Id,
            UserId = destination.UserId,
            Type = destination.Type,
            Title = destination.Title,
            Body = destination.Body,
            CreationDate = destination.CreationDate,
            WasUpdated = destination.WasUpdated,
            WasDeleted = destination.WasDeleted,
            CommentCount = destination.CommentCount,
            Upvotes = destination.Votes.Upvotes,
            Downvotes = destination.Votes.Downvotes,
            SpaceId = destination.SpaceId
        };
        #endregion
    }
}